### PR TITLE
feat: logger

### PR DESCRIPTION
**Pull request by bartoszherba**
closing https://alokai.atlassian.net/jira/polaris/projects/PROD/ideas/view/1113287?selectedIssue=PROD-91

### ❓ Type of change

- [ ] 📖 Documentation (updates to the documentation, readme or JSDoc annotations)
- [ ] 🐞 Bug fix (a non-breaking change that fixes an issue)
- [ ] 👌 Enhancement (improving an existing functionality like performance)
- [x] ✨ New feature (a non-breaking change that adds functionality)
- [ ] 🧹 Chore (updates to the build process or auxiliary tools and libraries)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)

### 📝 Checklist

- [ ] I have linked an issue or discussion.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: .eslintrc.js
Comment (WojtekTheWebDev) [inline .eslintrc.js:4]: This is good enough, but I think that we could add it to the `@vue-storefront/eslint-config-integrations` package directly

File: packages/logger/src/LoggerFactory.ts
Comment (WojtekTheWebDev) [inline packages/logger/src/LoggerFactory.ts:9]: Can we assume that we have an opinion about the default logger and use it if someone won't pass the type?

**Review by WojtekTheWebDev — COMMENTED** [review]
eslint can be ignored but we could start a discussion about the default factory result

File: packages/logger/src/LoggerFactory.ts
Comment (dawid-ziobro) [inline packages/logger/src/LoggerFactory.ts:19]: Maybe we should merge default options with those that are passed through params?

Comment (bartoszherba) [reply packages/logger/src/LoggerFactory.ts]: ok, fixed

File: packages/logger/__tests__/unit/consolaStructuredLogger.spec.ts
Comment (Fifciu) [inline packages/logger/__tests__/unit/consolaStructuredLogger.spec.ts:15]: nit pick: I usually find `should` in the description of tests redundant. `It creates`, It logs` would be enough

File: packages/logger/__tests__/unit/loggerFactory.spec.ts
Comment (Fifciu) [inline packages/logger/__tests__/unit/loggerFactory.spec.ts:8]: It doesn't really check if consola-gcp type has been created

File: packages/logger/__tests__/unit/structuredLog/gcpStructuredLog.spec.ts
Comment (Fifciu) [inline packages/logger/__tests__/unit/structuredLog/gcpStructuredLog.spec.ts:153]: I get the idea of using .each. However, I always find it overwhelming, especially here when we have 10 cases. This test was much harder to grasp than others. Imagine looking at it at midnight when you have to fix smth in a hurry, or just 6 months from now. Please consider refactoring this test to a few granular ones with a detailed description.

File: packages/logger/src/ConsolaStructuredLogger.ts
Comment (Fifciu) [inline packages/logger/src/ConsolaStructuredLogger.ts:78]:
```suggestion
    // A few LogLevel's map to the same ConsolaLevel, so we had to operate on ConsolaLevel's to DRY
    const consolaLevel = mapToConsolaLevel(level);
```

This line and switch made me think a minute about "why". Maybe we could add comment or some human-readable object like (I prefer latter, what about you?):
```ts
const CONSOLA_LEVELS = {
  ERROR: 0,
  WARN: 1,
};

// then
case ERROR:
logger.error({ structuredLog: structuredLogEntry });
        break;
```

File: packages/logger/src/ConsolaStructuredLogger.ts
Comment (bartoszherba) [reply packages/logger/src/ConsolaStructuredLogger.ts:60]: ok, addressed both problems

File: packages/logger/src/structuredLog/GCPStructuredLog.ts
Comment (bartoszherba) [reply packages/logger/src/structuredLog/GCPStructuredLog.ts:43]: fixed

File: packages/logger/__tests__/unit/consolaStructuredLogger.spec.ts
Comment (bartoszherba) [reply packages/logger/__tests__/unit/consolaStructuredLogger.spec.ts:15]: ok

File: packages/logger/__tests__/unit/loggerFactory.spec.ts
Comment (bartoszherba) [reply packages/logger/__tests__/unit/loggerFactory.spec.ts:8]: I know, since i changed it from the class => function factory I do not know how to do it without some kind of metadata on the object which will be there only for this test.

File: packages/logger/__tests__/unit/structuredLog/gcpStructuredLog.spec.ts
Comment (bartoszherba) [reply packages/logger/__tests__/unit/structuredLog/gcpStructuredLog.spec.ts:153]: I think that the matrix for a multiple inputs for the same test is something that we should not fight with. I am very used to such tests. Imagine more input data. This is exactly what it is for and the only argument is that from your POV it is less readable. Please, create a ticket in the jest repo :)

File: packages/logger/src/ConsolaStructuredLogger.ts
Comment (bartoszherba) [reply packages/logger/src/ConsolaStructuredLogger.ts:78]: I have added a comment.

File: packages/logger/__tests__/unit/loggerFactory.spec.ts
Comment (Fifciu) [reply packages/logger/__tests__/unit/loggerFactory.spec.ts:8]: Remove it then

**Review by filrak — CHANGES_REQUESTED** [review]
Looking at the examples in the docs I think we should consider extracting logger via `this.logger.x` instead. More explanation in the comments

File: docs/content/3.middleware/2.guides/10.working-with-logger.md
Comment (michaelKurowski) [reply docs/content/3.middleware/2.guides/10.working-with-logger.md:98]: I did discuss it with Bartek, I honestly don't know why do we have to have this as a function and the argument that "well we may want to put it in a different place later" doesn't really speak to me. It should either be context or req.
Think about this that way - now user has to import it to even use it, rather than simply access is, so it increases the chances that the person will use `console` instead of our logger.

`this` would be a new convention in our product, so I'm against this.


I won't die on this hill tho, the logger was in a development already for quite some time.

Comment (filrak) [reply docs/content/3.middleware/2.guides/10.working-with-logger.md:98]: Agree

Comment (bartoszherba) [reply docs/content/3.middleware/2.guides/10.working-with-logger.md:98]: Firstly, do not use colorized sentences like "well we may want to put it in a different place later". You put us in a place where we do not have arguments, where it is not true. It is neither a quote, neither our position in this case. We have provided a logic arguments and you respond with what you think which is a word against word at best.

1. it is not possible to logger only on the context or req (req is unusable at all and we must use res as only res have locals to store some custom data).
2. in some places only context is available, in some only res
3. in some places, like hooks, which is a great example for you, we can't use neither of them because of race conditions in hooks execution. We can't inject logger into singletons.
4. we have discussion with DX about the approach and we showed alternatives
5. logger is set on:

- res.locals.alokai.logger
- context.logger
- someObject.res
- container.logger (injected into hooks to avoid race conditions issues)

So the user will have to remember all these use cases to be able to get the logger.
Also, if we change the path, it will be a breaking change for all users and for all our packages. 

To address these issues we introduced getLogger function that hide all that complexity which which, in addition, prevents from bc.
Even now your request implies changes in 20 packages at least, and only 20 because we did not add logger everywhere yet.

Comment (michaelKurowski) [reply docs/content/3.middleware/2.guides/10.working-with-logger.md:98]: I wasn't attempting a strawman, sorry if it looked this way. This was the final argument I remembered from our call after the conversation we've had. Not exactly in this phrasing but the summary of it.
- The fact that req is unusable I don't remember at all, if you said it it wasn't back on our calls. 
- Putting it on res is something completely new to me.
- We already had a discussion that there don't have to be any singletons, we have a full control over what do we pass to parameters of hooks and methods, we can build objects on the fly.

That being said, it doesn't really matter at the end of the day. We would go back and argue in circles again and it is greenlighted already.

Comment (bartoszherba) [reply docs/content/3.middleware/2.guides/10.working-with-logger.md:98]: `The fact that req is unusable I don't remember at all, if you said it it wasn't back on our calls.` 
You can't remember that because we did not know that at that time.
`We already had a discussion that there don't have to be any singletons, we have a full control over what do we pass to parameters of hooks and methods, we can build objects on the fly.`
But the req/res object is and will be a singleton - and must be a singleton. And the same req/res is used in parallel. We have control of what we pass but we can't pass it into these object in all places (eg. in hooks, we just cant due to condition race).

We have considered `this.logger` but it is far from being a perfect solution. It will not work for arrow functions which is the major of our code, we can't bind this. Also, using `this` in such way is an antipattern.

We can also consider attaching getLogger to the global object - which is also an anti pattern but, at least, it is not "this".

Comment (Fifciu) [reply docs/content/3.middleware/2.guides/10.working-with-logger.md:98]: @filrak Indeed, it would be pleasant to have it always available via `this`. However, we cannot bind `this` for arrow functions. Example (last 2 lines are most important):
```ts
const obj = {
  value: 'Object',
  regularFunction: function() {
    console.log(this.value);
  },
  arrowFunction: () => {
    console.log(this.value);
  }
};

obj.regularFunction(); // Will print 'Object'
obj.arrowFunction(); // Will print 'undefined' in the browser or error in Node.js

// New binding of `this` with `bind`
const newRegularFunction = obj.regularFunction.bind({ value: 'New Object' });
newRegularFunction(); // Will print 'New Object'

const newArrowFunction = obj.arrowFunction.bind({ value: 'New Object' });
newArrowFunction(); // Will still print 'undefined' in the browser or error in Node.js

```

So this change would require rewriting every handler using logger in every integration/module/extension to be classic non-arrow function. We would also need to have constraint to only use classic functions in some places of our software. 

What's more, it doesn't solve the problem of accessing the wrong logger via closure, e.g:
```ts
const someExtension = {
  name: "some-extension",
  hooks(req, res, alokai) {
    // const hooksLogger = getLogger(alokai);
    const hooksLogger = this.logger;
    hooksLogger.info("hooks");
    return {
      beforeCall(params) {
        // Never access via closure a logger belonging to the hooks factory:
        hooksLogger.info("smth");
        // Instead use own logger of every hook function:
        const logger = this.logger;
        // const logger = getLogger(params);
        logger.info("smth");
        return params.args;
      }
    };
  },
};
```

Which unfortunately, invalidates an idea of passing logger via `this`.

BTW I don't understand mentioning LoggerManager. It's not public thing anymore, it's used only internally. So end-developer won't be overwhelmed by new thing to learn.

@michaelKurowski Yup, we learned a lot more during work on the logger and haven't had a chance yet to tell about the details of our findings. 

> I wasn’t aware that there are places exposed to the user that only have res available, looking at the code now, looks like you probably refer to the after hook?

Bartek used mental shortcut, he meant we have access to either context or req+res pair.

BTW At first, I also thought `after hook` has access to the `response: Response` from express. However, it has value of response.body. 

### Our final approach 
It mostly involves injecting a logger wrapped with proper metadata in every possible place. Every place has its own set of metadata attached to make it resistant to singleton-related issues and race conditions. We couldn't set it in `context.res` and attach metadata on the fly, as it causes issues in case of [parallel calls of multiple handlers in the scope of one request](https://docs.alokai.com/middleware/guides/orchestration#example-1-fetching-custom-product-properties-from-legacy-systems) and would make code really hard to grasp.

I didn't want to be too overwhelming with explanation so I've cut to the chase. If you wish to know more about challenges, tell me, as I am open to writing additional longer comment.

File: docs/content/3.middleware/2.guides/10.working-with-logger.md
Comment (Fifciu) [reply docs/content/3.middleware/2.guides/10.working-with-logger.md:47]: Great idea!

File: docs/content/3.middleware/2.guides/10.working-with-logger.md
Comment (filrak) [reply docs/content/3.middleware/2.guides/10.working-with-logger.md]: Can you elaborate more on this final approach? 

I get the argument to not use `this` and indeed one less import is not worth refactoring all packages. Though If we can still have it available directly from the function parameters instead of relying on `getLogger` import and having possibility to use it wrong when we don't inject it on every level would be a step forward in reducing complexity. Can we have that?


Also" it's used only internally. So end-developer won't be overwhelmed by new thing to learn.". Wrong. Integration API is also a public API. Customers often write their own integrations with various systems.

Comment (bartoszherba) [reply docs/content/3.middleware/2.guides/10.working-with-logger.md]: I will let @Fifciu to answer it but:
`Though If we can still have it available directly from the function parameters instead of relying on getLogger`
the logger itself is in very different places depending on if this is a hook, extension, or other extensibility place. End user will have to remember where and how to get it for all these places. `getLogger` hides this complexity.
`and having possibility to use it wrong when we don't inject it on every level would be a step forward in reducing complexity.` I do not understand the rest of it; What possibility to use it wrong? Where we do not inject the logger? Could you provide some examples where `getLogger` leads to issues and function-less approach does not?

As I described, here are 3 places where logger can be accessed, depends on the extension point:

res.locals.alokai.logger - if res if available and we do not have to care about race condition
context.logger - if context is available but res is not
alokai.logger - injected into hooks to avoid race conditions issues

`getLogger` not only hides that complexity, protect clients from bc but also help with types inference - it would be a hard or impossible to rework broken typing in middleware in such way that the logger will be properly typed in all these places.

`Also" it's used only internally. So end-developer won't be overwhelmed by new thing to learn.". Wrong. Integration API is also a public API. Customers often write their own integrations with various systems.` This is about something completely different, Filip was referring to some internal things which are irrelevant here. We still use loggerManager internally.

Comment (bartoszherba) [reply docs/content/3.middleware/2.guides/10.working-with-logger.md]: @filrak This is our final list of arguments, if you do not think that these are valid points we will do as you ask.

- Users do not need to remember where to obtain the logger, as it is encapsulated within the function itself.
- The function can provide documentation within a docblock, immediately clarifying the objects it operates on (beyond TypeScript types). This eliminates the need to leave the IDE to refer to external documentation.
- By encapsulating specific logic within the function, we make it easier to modify in the future and protect clients from breaking changes.
- This approach enables us to maintain proper typing, which might not be achievable in our middleware under other circumstances (to be validated).
- Every method of accessing the logger must be documented. When using a function, the docblock and brief in-documentation note provide clear information, ensuring comprehensive documentation.

Comment (bartoszherba) [reply docs/content/3.middleware/2.guides/10.working-with-logger.md]: @filrak @michaelKurowski 
Guys, what is your final decision? Can we use `getLogger` or we should get rid of it from the public interface?

Comment (filrak) [reply docs/content/3.middleware/2.guides/10.working-with-logger.md]: Thanks for giving the arguments, they are convincing.

Frankly I think given your arguments choosing one or the other will result in tradeoffs coming purely from the weakness of our implementation rather than thinking what is the best possible API for this kind of feature, and I'm sure there is an approach that won't have them (some form of context-aware `getLogger` from fn context/params ). Though, if you don't see such solution at first glance lets proceed with your proposal.

File: docs/content/3.middleware/2.guides/10.logging.md
Comment (filrak) [inline docs/content/3.middleware/2.guides/10.logging.md:240]:
```suggestion
    "level": "error"
```
one last catch, wouldn't `verbosity` be more descriptive as we're talking specifically about log verbosity levels here/

**Review by filrak — CHANGES_REQUESTED** [review]

**Review by filrak — APPROVED** [review]

File: packages/logger/src/interfaces/LoggerInterface.ts
Comment (Fifciu) [inline packages/logger/src/interfaces/LoggerInterface.ts:12]: We've never used it eventually. Let's remove it and add later if needed

**Review by Fifciu — CHANGES_REQUESTED** [review]

File: docs/content/3.middleware/2.guides/10.logging.md
Comment (dawid-ziobro) [inline docs/content/3.middleware/2.guides/10.logging.md:315]: Example should be wrapped with 
```
export const config = {
```

File: docs/content/3.middleware/2.guides/10.logging.md
Comment (dawid-ziobro) [inline docs/content/3.middleware/2.guides/10.logging.md:337]: Also here should be wrapped with 
```
export const config = {
```

File: docs/content/3.middleware/2.guides/10.logging.md
Comment (dawid-ziobro) [inline docs/content/3.middleware/2.guides/10.logging.md:11]:
```suggestion
The examples below demonstrate how to use the application logger to ensure that correct metadata is conveyed in the logs. To achieve that, we will use the `getLogger` function to extract the logger instance from the passed argument.
```

File: docs/content/3.middleware/2.guides/10.logging.md
Comment (dawid-ziobro) [inline docs/content/3.middleware/2.guides/10.logging.md:224]:
```suggestion
### Init
```

File: docs/content/3.middleware/2.guides/10.logging.md
Comment (bartoszherba) [reply docs/content/3.middleware/2.guides/10.logging.md:315]: ok, fixed

File: docs/content/3.middleware/2.guides/10.logging.md
Comment (bartoszherba) [reply docs/content/3.middleware/2.guides/10.logging.md:337]: fixed

File: docs/content/3.middleware/2.guides/10.logging.md
Comment (bartoszherba) [reply docs/content/3.middleware/2.guides/10.logging.md:11]: fixed

File: docs/content/3.middleware/2.guides/10.logging.md
Comment (dawid-ziobro) [inline docs/content/3.middleware/2.guides/10.logging.md:64]:
```suggestion
      // ...
```

File: docs/content/3.middleware/2.guides/10.logging.md
Comment (dawid-ziobro) [inline docs/content/3.middleware/2.guides/10.logging.md:39]:
```suggestion
      // ...
```

File: packages/logger/__tests__/unit/structuredLog/gcpStructuredLog.spec.ts
Comment (dawid-ziobro) [inline packages/logger/__tests__/unit/structuredLog/gcpStructuredLog.spec.ts:6]: Shouldn't it be `development`?

File: packages/middleware/src/integrations/helpers/resolveDependency.ts
Comment (dawid-ziobro) [inline packages/middleware/src/integrations/helpers/resolveDependency.ts:1]:
```suggestion
import type { AlokaiContainer } from "../../types";
```

File: packages/middleware/src/logger/injectMetadata.ts
Comment (dawid-ziobro) [inline packages/middleware/src/logger/injectMetadata.ts:3]:
```suggestion
import type { AlokaiLocal } from "../types";
```

File: packages/middleware/src/types/common.ts
Comment (dawid-ziobro) [inline packages/middleware/src/types/common.ts:2]:
```suggestion
import type { LoggerInterface } from "@vue-storefront/logger";
```

File: packages/middleware/src/types/common.ts
Comment (dawid-ziobro) [inline packages/middleware/src/types/common.ts:3]:
```suggestion
import type { LoggerOptions } from "./config";
```

File: packages/middleware/src/apiClientFactory/wrapFnWithErrorBoundary.ts
Comment (dawid-ziobro) [inline packages/middleware/src/apiClientFactory/wrapFnWithErrorBoundary.ts:1]:
```suggestion
import type { LogScope } from "../types";
```

**Review by dawid-ziobro — APPROVED** [review]

**Review by Fifciu — APPROVED** [review]

File: .changeset/blue-suits-drop.md
Comment (WojtekTheWebDev) [inline .changeset/blue-suits-drop.md:5]: All those changesets messages will appear in the final changelog, you should consolidate them into a single one or just remove those with messages like `rc`.

File: packages/logger/CHANGELOG.md
Comment (WojtekTheWebDev) [inline packages/logger/CHANGELOG.md:3]: Imho no point of having those messages in here as well, before merging into main let's clean it up.

File: packages/logger/package.json
Comment (WojtekTheWebDev) [inline packages/logger/package.json:28]: Do you need `dotenv` dependnecy in here? Shouldn't it be a devDep or peerDep?

File: packages/middleware/CHANGELOG.md
Comment (WojtekTheWebDev) [inline packages/middleware/CHANGELOG.md:3]: Another changelog to cleanup

**Review by WojtekTheWebDev — APPROVED** [review]
Minor comments to consider before merging from my side

File: packages/logger/__tests__/unit/structuredLog/gcpStructuredLog.spec.ts
Comment (bartoszherba) [reply packages/logger/__tests__/unit/structuredLog/gcpStructuredLog.spec.ts:6]: it should, fixed

**Comment by Tofandel** [comment]
Since this change I get a `Failed to resolve apiClient or function: Logger instance could not be determined` on all endpoints

**Pull request merged**
